rfd/0085-better-react-testing-remove-default-exports.md

```diff
@@ -0,0 +1,155 @@
+---
+authors: Ryan Clark (ryan.clark@goteleport.com)
+state: draft
+---
+
+# RFD 85 - Better React testing, remove default exports
+
+## What
+
+Improve the webapps codebase by changing the way we test our components and write our component stories.
```

Comment:
i'm not sure if this RFD proposes a different way to write our component stories. are we keeping the smart/dummy approach? (keeping it was my assumption)

Reply:
We can keep it if wanted, but it should focus more on displaying individual components instead of a "full story".

For example, I have `Component` that loads data and returns either `ComponentLoadingFailed`, `ComponentContent` or `ComponentLoading`. You can write stories for these sub-components, instead of the main `Component`.

This is to stop all of the state for `Component` being passed in as props, and then us adding a wrapper component that grabs all of this stuff from a hook. It makes things extremely hard to track down.

Instead, all logic for fetching data can be in `Component`, and then the relevant data and callbacks are passed into the children components, keeping their props to only what they need.

Reply:
> For example, I have Component that loads data and returns either ComponentLoadingFailed, ComponentContent or ComponentLoading. You can write stories for these sub-components, instead of the main Component

Usually it's better to have _both when needed._ The full story lets you see more complete view when making changes (specifically to layout). Also most of the sub-components do not need a story _at all_ as that's the way to split up HTML.

> And then us adding a wrapper component that grabs all of this stuff from a hook. It makes things extremely hard to track down.

That's the common pattern (used for Containers) and I am not sure why it became `extremely hard to track down` as we had interns working on those components and got positive feedback.

> Instead, all logic for fetching data can be in Component, and then the relevant data and callbacks are passed into the children components, keeping their props to only what they need.

This comment contradicts with the `extremely hard to track down` as View is a child of Container too. Also, by placing API code to the Component itself it will make it very hard to snapshot a story (because state changes dynamically). It will also make it hard to write stories as one would need to write code to bring a view to a particular state. It will also bring issues related to mocks done in storybook (cleanups). Most likely devs will stop using storybook _at all_ and will have to run a cluster for any UI work. Different states will be burred in the code with no easy way to visually track them. CSS regression bugs will start eventually popping up in bigger numbers. And the only stories to be left is for basic components (design system) which is what industry average does.

Just to summarize.

1. We do not have an [industry average problem](https://medium.com/@sapegin/whats-wrong-with-snapshot-tests-37fbe20dfe8e) where devs misuse snapshots. Every snapshot has a visual story and never used to _test_ behavior.
2. Add a behavior test if missing.
3. Storybook is a _generic tool_ that can be used for different purposes.
4. Remove default exports.

At the end, it's all about being productive, move faster, and better dev UX. I was not able to see how this RFD improves addresses any of it but if the team sees it differently then go for it :)

Reply:
I think @alex-kovoy makes an important point in that if we move away from this containerized approach, UI development becomes significantly more difficult in that we either need to build a behavioral test that induces the state we want to try out, or else run an entire cluster and then manually induce that state.

There's nothing about our current approach that prevents us from _also_ writing behavioral tests, whereas if we elect to remove the containers then we won't be able to easily create storybooks.

Reply:
Basically the second paragraph in @alex-kovoy's other comment [here](https://github.com/gravitational/teleport/pull/15813/files/368b385a3752c205db41fb93837089b768b1f710#r1019409689)

Reply:
We could also move behavior tests to the storybook. https://storybook.js.org/docs/react/writing-tests/interaction-testing 
The implementation would look like this:
1. mock network requests eg using `msw`
2. Render component
3. Define interactions that will trigger the desired state.
4. assert correct view state

You can also check the results in the storybook. We can run these tests in CI pipeline https://storybook.js.org/docs/react/writing-tests/test-runner#run-against-non-deployed-storybooks

Reply:
I understand the general idea, the point that @alex-kovoy made and that I'm reiterating is that 1, 2, and 3 all take considerably more work than just passing in the desired state via props. 

In a sense what we're doing currently is allowing a component's internal state to be "`public`" fields, whereas without containers, most of a component's state would become "`private`" fields. It's a controversial topic, but I'm of the opinion that directly manipulating what are effectively `"private`" fields is useful for various types of tests and should be allowed by language semantics. See [here](https://stackoverflow.com/a/6778726) for what some do in C++ to achieve this, and note that languages like Go and Rust acknowledge this by allowing "`private`" fields to be accessible within packages/modules (where unit tests are written).

There's nothing about the container pattern that prevents us from writing interaction tests, but if we eschew the container then we will no longer be able to quickly and easily iterate on UI components that have complex internal states. Instead we will be forced to write more involved rube-goldberg machines of mocked network requests and UI interactions to achieve what's currently trivial (just hardcode the props to whatever state you want).

I can see the argument being made that actually that's a good thing to force us to do, because that ensures that we write these more substantive interaction tests which otherwise we won't write because we'll have an easy way out. However given that I'm not aware of a particular problem with behavioral regressions, I don't think force-slowing-down development is warranted.

I suppose @ryanclark might argue that actually the bigger slowdown is the tracking down of these props. Granted that it is more of a process than if most internal state where right there in the component, however I noticed that over time I've gotten better at tracking down props. Additionally, over time I've come to rely more on storybook for development.

Reply:
I think this discussion could benefit greatly if we take typical complex feature component tests/stories and write them in a new way.

Reply:
A component I have in mind specifically is the `DesktopSession` and [its accompanying Storybooks](https://github.com/gravitational/webapps/blob/master/packages/teleport/src/DesktopSession/DesktopSession.story.tsx). It's a very complex component -- multiple fetches, a websocket that feeds into a client which emits events that are listened to and the UI is changed in response. I think it would be very difficult to mock.

Granted that's perhaps an exception... If you (or anybody) can create an example of a more typical component with private state, whose state is easy to manipulate with mocked calls and interactions, I'm willing to use it as a template and try it out on another component and see if it changes my mind.

I'll note that there's nothing set in stone that we have to go with one approach or the other, we can make component-by-component judgement calls. The main reason I'm pushing back is that this proposes to do away with the state-as-props pattern altogether, which I think would slow me down significantly on i.e. `DesktopSession`.

Reply:
> The difference here is that we can also run and inspect this test in the storybook. No need for testing the state and creating a story for the state separately. 

@JanKaczmarkiewicz  Storybook is not a great environment for mocking. Use jest for that. And vice versa. 

Container pattern is very common in the industry. If it helps, think of it as _an explicit_ version of Redux `connect` where state and actions are mapped to props.

Before we try to simplify `JSON`, I suggest to take a look [at this component](https://github.com/gravitational/webapps/tree/master/packages/teleport/src/Discover) written without any stories. Not only it provides no instant clue of what it does, it's also an engineering resource drain for any devs to make a minor CSS change.

Again, treat stories as first class citizens and this alone will answer _why_ things at Teleport are done this way. Be careful with blindly applying vanity metrics and respect and value Dev UX (i.e. company resources)